pkg/codegen/schema/schema.go

```diff
@@ -925,63 +941,150 @@ func (t *types) bindPrimitiveType(name string) (Type, error) {
 	}
 }
 
-func (t *types) bindType(spec TypeSpec) (Type, error) {
-	if spec.Ref != "" {
-		switch spec.Ref {
-		case "pulumi.json#/Archive":
-			return ArchiveType, nil
-		case "pulumi.json#/Asset":
-			return AssetType, nil
-		case "pulumi.json#/Json":
-			return JSONType, nil
-		case "pulumi.json#/Any":
-			return AnyType, nil
+// externalSchemaRef contains the parsed fields for an external schema ref.
+type externalSchemaRef struct {
+	Package string          // The package component of the schema ref
+	Version *semver.Version // The version component of the schema ref
+}
+
+// typeSpecRef contains the parsed fields from a type spec reference.
+type typeSpecRef struct {
+	*externalSchemaRef
+
+	URL   *url.URL // The parsed URL
+	Token string   // The type token
+	Kind  string   // The kind of reference, either 'resources' or 'types'
+}
+
+// Regexs used to parse type spec refs. These are declared at the package scope to avoid repeatedly recompiling them.
+var refPathRegex = regexp.MustCompile(`^/(?P<package>\w+)/v?(?P<version>\d+\.\d+\.\d+)/schema\.json$`)
+var refFragmentRegex = regexp.MustCompile(`^/(?P<kind>resources|types)/(?P<token>.*:.*:.*)$`)
+
+func parseTypeSpecRef(ref string) typeSpecRef {
+	parsedURL, err := url.Parse(ref)
+	contract.AssertNoErrorf(err, "failed to parse ref URL: %s", ref)
```

Comment:
I would rather these be errors than assertions: these could arise from malformed user input.

Reply:
(same throughout this function)